File: jvm/src/main/kotlin/com/jetbrains/signatureverifier/serialization/dataholders/attributeInfos/SigningTimeAttributeInfo.kt
package com.jetbrains.signatureverifier.serialization.dataholders

import com.jetbrains.signatureverifier.serialization.OffsetDateTimeSerializer
import com.jetbrains.signatureverifier.serialization.toDLSet
import kotlinx.serialization.Serializable
import org.bouncycastle.asn1.ASN1UTCTime
import org.bouncycastle.asn1.cms.Attribute
import java.time.OffsetDateTime
import java.time.format.DateTimeFormatter

@Serializable
class SigningTimeAttributeInfo(
  override val identifier: TextualInfo,
  val content: List<@Serializable(OffsetDateTimeSerializer::class) OffsetDateTime>
) : AttributeInfo() {
  constructor(attribute: Attribute) : this(
    TextualInfo.getInstance(attribute.attrType),
    attribute.attributeValues.map {
      OffsetDateTime.parse(
        it.toString(),
        dateTimeFormatter
      )
    }
  )

  companion object {
    val dateTimeFormatter: DateTimeFormatter = DateTimeFormatter.ofPattern("yyMMddHHmmssX")
  }

  override fun getPrimitiveContent() = content.map {
    ASN1UTCTime(
      it.format(dateTimeFormatter)
    )
  }.toDLSet()
}

File: net/JetBrains.Serialization/src/Dataholders/AttributeInfos/SigningTimeAttributeInfo.cs
using Newtonsoft.Json;
using Org.BouncyCastle.Asn1;
using Attribute = Org.BouncyCastle.Asn1.Cms.Attribute;
using JetBrains.Serialization;
using System.Globalization;

[JsonObject(MemberSerialization.OptIn)]
public class SigningTimeAttributeInfo : AttributeInfo
{
  [JsonProperty("Identifier")] protected override TextualInfo Identifier { get; }

  [JsonProperty("Content")] private List<DateTimeOffset> _content;
  private static readonly string DateTimeFormat = "yyMMddHHmmssZ";

  [JsonConstructor]
  public SigningTimeAttributeInfo(TextualInfo identifier, List<DateTimeOffset> content)
  {
    Identifier = identifier;
    _content = content;
  }

  public SigningTimeAttributeInfo(Attribute attribute)
  {
    Identifier = TextualInfo.GetInstance(attribute.AttrType);
    _content = attribute.AttrValues.ToArray().Select(item =>
      DateTimeOffset.ParseExact(item.ToString(), DateTimeFormat, CultureInfo.InvariantCulture)).ToList();
  }

  public override Asn1Encodable GetPrimitiveContent() =>
    _content.Select(time => new DerUtcTime(time.ToString(DateTimeFormat))).ToDerSet();
}